### PR TITLE
BaseModel

### PR DESCRIPTION
**Pull request by Grinch101**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,7 @@ __pycache__
 *.pyc
 A
 templates/debug.log
+contact.py
```
Comment (martianboy) [inline .gitignore]: You either delete files or let them be. Ignoring a code file is really weird!

File: MyModels.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+class BaseModel:
```
Comment (martianboy) [inline MyModels.py]: Please put these models back into their own separate files.

File: MyModels.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(self):
+        self.list = []
+        self.id_index = {}
```
Comment (martianboy) [inline MyModels.py]: This index is not updated by the BaseModel itself. Let's move that logic inside, as it's the most straightforward compared to other indexes.

File: MyModels.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    def find_index(self,id):
+        return self.id_index[id]
+    
+    def find_val(self,id):
+        index = self.find_index(id)
+        return self.list[index]
+
+
+###################
+    def GetAll(self):
```
Comment (martianboy) [inline MyModels.py]: Your naming need to be consistent. CamelCasing one function while snake_casing others is gonna look weird. In Python, CamelCased names are for classes, snake_cased names are for methods and variables.

File: MyModels.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        return self.list
+    
+    def delete(self, entry):
+        
+        self.list.remove(entry)
+    
+    def update(self,id, new_entry):
```
Comment (martianboy) [inline MyModels.py]: Make sure you format your code. Here, there should be a space after the first comma: `self, id, ...`

File: MyModels.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    def find_val(self,id):
+        index = self.find_index(id)
+        return self.list[index]
+
+
+###################
+    def GetAll(self):
+        return self.list
+    
+    def delete(self, entry):
```
Comment (martianboy) [inline MyModels.py]: Usually when you wanna delete an entry, you only have its `id`, not a memory reference to its full entry. Therefore, a `delete` method that takes an id for deleting is much more desirable than this.

File: MyModels.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    def register(self, client_name, email, password):
+        entry = {'client_name' : client_name,
+                'email': email,
+                'password':password,
+                'userid': self.userid }
+        self.add(entry = entry)
+        self.id_index[self.userid] = len(self.list) - 1
+        self.email_userid[email] = self.userid
+        
+        self.userid += 1
```
Comment (martianboy) [inline MyModels.py]: You were supposed to rename this to `id_sequence_counter`, or `id_seq_counter` if you prefer shorter names. Also this action is common between all models and needs to go up into the `BaseModel`.

File: MyModels.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        self.email_userid = {}
+        self.userid = 0
+        super().__init__()
+        
+    
+    def register(self, client_name, email, password):
+        entry = {'client_name' : client_name,
+                'email': email,
+                'password':password,
+                'userid': self.userid }
```
Comment (martianboy) [inline MyModels.py]: `userid` field should be renamed to `id`. Always name the main primary key as `id` unless you have a really good reason not to.

Comment (martianboy) [reply MyModels.py]: Also note that assigning the `id` field is common between models. See also line 122 for the similar code in the Contact model. Something tells me that it should not be a responsibility of child model classes.

File: MyModels.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        
+    def delete(self,userid):
+        
+        entry = self.find_val(userid)
+        super().delete(entry)
+        del self.id_index[userid]
+        for userid in range(userid+1 , len(self.list)+1):
+            self.id_index[userid] -= 1
+        
+    def update(self,userid,new_value,entity = 'email'):
```
Comment (martianboy) [inline MyModels.py]: I don't get what this function does at all! What does the last argument, `entity`, do?

**Review by martianboy — CHANGES_REQUESTED** [review]

File: MyModels.py (hunk 2 above)
Comment (Grinch101) [reply MyModels.py]: Sure, will do.

File: MyModels.py (hunk 3 above)
Comment (Grinch101) [reply MyModels.py]: You are right! this situation has been created after editing code and simplifying the basemodel

File: MyModels.py (hunk 5 above)
Comment (Grinch101) [reply MyModels.py]: Sure, Thanks.

File: MyModels.py (hunk 6 above)
Comment (Grinch101) [reply MyModels.py]: same problem due to simplifying ! I assure you BaseModel will have great changes in the very next commit.

File: MyModels.py (hunk 7 above)
Comment (Grinch101) [reply MyModels.py]: Now, a question arises here, I will ask it in Telegram.

File: MyModels.py (hunk 9 above)
Comment (Grinch101) [reply MyModels.py]: it updates a key/value pair of  the inserted entry! entity tells us which key/value pair we are going to edit

File: MyModels/BaseModel.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self):
+        self.list = []
+        self.id_index = {}
+        self.id_sequence_counter = 0
+        
+    def add(self, entry):
+        
+        self.list.append(entry)
+
+        userid = entry['userid'] # Applicable in Contact.py Only
```
Comment (martianboy) [inline MyModels/BaseModel.py]: `BaseModel` cannot have any child-specific logic. It cannot know anything about the existence of a `user_id` or anything. This whole block should be taken out.

File: app.py [hunk 11]
```diff
@@ -1,7 +1,6 @@
 from flask import Flask, render_template, request, flash, redirect, url_for, make_response
-# from contact import Contact
-# from user import User
-from MyModels import Contact, User
+from MyModels.ContactFolder.contact import Contact
+from MyModels.UserFolder.user import User
```
Comment (martianboy) [inline app.py]: See how ugly these look? `UserFolder` in a module import.

File: MyModels/BaseModel.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+                self.userid_id[userid].append(self.id_sequence_counter)
+            else:
+                self.userid_id[userid] = [self.id_sequence_counter]
+            
+        self.id_index[self.id_sequence_counter] = len(self.list) - 1
+        
+        self.id_sequence_counter += 1
+
+
+    def find_index(self,id_sequence_counter):
```
Comment (martianboy) [inline MyModels/BaseModel.py]: ????? What does `id_sequence_counter` have to do here?

File: MyModels/BaseModel.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        return self.id_index[id_sequence_counter]
+    
+    def find_val(self,id_sequence_counter):
```
Comment (martianboy) [inline MyModels/BaseModel.py]: And here?

File: MyModels/BaseModel.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        index = self.find_index(id_sequence_counter)
+        return self.list[index]
+
+    def get_all(self):
+        return self.list
+    
+    def delete(self, id_sequence_counter):
```
Comment (martianboy) [inline MyModels/BaseModel.py]: And here?

File: MyModels/BaseModel.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        
+        entry = self.find_val(id_sequence_counter)
+        self.list.remove(entry)
+        del entry
+    
+    def update(self, id_sequence_counter, new_entry):
```
Comment (martianboy) [inline MyModels/BaseModel.py]: And here?

**Review by martianboy — CHANGES_REQUESTED** [review]

File: MyModels/ContactFolder/contact.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    def insert(self, userid, client_name, name, phone):
+
+        entry = {'userid': userid,
+                 'client_name': client_name,
+                 'name': name,
+                 'phone': phone,
+                 'id': self.id_sequence_counter}
+        self.add(entry)
+
+    def delete(self, id_sequence_counter):
```
Comment (martianboy) [inline MyModels/ContactFolder/contact.py]: Baba you've completely confused `id` with `id_seq_counter`!

File: MyModels/BaseModel.py (hunk 12 above)
Comment (Grinch101) [reply MyModels/BaseModel.py]: for the sake of readability! did it backfired?!

**Pull request merged**